lib/modules.nix

```diff
@@ -110,8 +79,12 @@ rec {
                   # there's _module.args. If specialArgs.modulesPath is defined it will be
                   # used as the base path for disabledModules.
                   specialArgs ? {}
-                , # This would be remove in the future, Prefer _module.args option instead.
-                  args ? {}
+                , # `class`:
+                  # A nominal type for modules. When set and non-null, this adds a check to
+                  # make sure that only compatible modules are imported.
+                  # This would be remove in the future, Prefer _module.args option instead.
+                  class ? null
```

Comment:
How about changing this to `expectedClass` (and `_module.expectedClass`), in order to distinguish it from the special `_class` per-module attribute, otherwise these are really easy to confuse.

Reply:
Although we've scoped it out for this PR, this won't be quite as applicable when `types.configuration` is added. In that use case `expectedClass` would have the role of the actual class in the check comparison.
Like a type, I see it as more of an intrinsic property than a detail that serves a single check.

Reply:
I'm not sure I get the connection to `types.configuration`, can you elaborate? I also don't yet know why `types.configuration` would be needed.

Reply:
`evalModules` takes modules with a `_class` and checks it against `_module.class`.
`types.configuration` takes `_module.class` from the `evalModules` return value and checks it against the type's `class`.

> why `types.configuration` would be needed

Submodule is generally preferable, but `evalModules` results do occur in flake-parts; for example [`nixosConfigurations`](https://flake.parts/options/flake-parts.html#opt-flake.nixosConfigurations).

I wouldn't mind defining the type elsewhere, but `lib.types` is a natural place for it.
Crucially though, the `(evalModules _)._module.class` is an "actual" class, and is not used as an expected class anymore, regardless of `types.configuration`.

Reply:
Hmm I see, that makes sense.

However I think `_module.*` should only be treated as inputs to the module system evaluation, not outputs. It also doesn't make sense for `_module.args` or `_module.check` to be a property of an evaluated configuration, at most to inspect/replicate how evaluation happened.

How about this: Call the argument and option `expectedClass`, but additionally expose it in the result as `(evalModules ...).class`. Or maybe even better: `.configurationClass`, relating more closely to `._type == "configuration"`.

Reply:
I've implemented the `(evalModules _).class` idea, but reverted the rename. Having to explain that these `*Class` names all refer to the same thing is not so great. The role of the class variable can be inferred from context, and since they have to be equal anyway, it doesn't even matter much which is which. Adding qualifiers in front of it unnecessarily eats brain space.
The `configuration` qualifier (in `configurationClass`) is implied by the `_type`, so I don't think we need to repeat it. It's unlikely to be ambiguous with the domain, and unlikely to conflict with a potential other framework, which should probably also use `_` prefixes for its framework-specific fields.